File: php/kirishima.php
<?php
    $numOfCells = (int)fgets(STDIN);
    $goal = $numOfCells - 1;
    $cells = split(" ", trim(fgets(STDIN)));

    const CURRENT = 0;
    const FINISHED = 1;
    const DEAD = 3;

    $acts = (int)fgets(STDIN);
    for ($act = 0; $act < $acts; $act++) {
        $actNum = (int)fgets(STDIN);
        $moveLog = array();
        $status = CURRENT;
        while ($status == CURRENT) {
            if ($actNum == $goal) {
                $status = FINISHED;
            }elseif($actNum < 1 || $goal < $actNum) {
                $status = DEAD;
            }else{
                $moves = (int)$cells[$actNum];
                if ($moves == 0) {
                    $status = DEAD;
                }else{
                    $actNum += $moves;
                    foreach ($moveLog as $movedNum) {
                        if ($actNum == $movedNum) {
                            $status = DEAD;
                            break;
                        }
                    }
                    if ($status != DEAD) {
                        $moveLog[] = $actNum;
                    }
                }
            }
        }
        print($status == FINISHED ? "Yes\n" : "No\n");
    }
?>

File: perl/kirishima.pl
use strict;
use warnings;

use constant {
    CURRENT => 0,
    FINISHED => 1,
    DEAD => 2,
};

my $str = <STDIN>;
chomp($str);
my $numOfCells = $str + 0;
my $goal = $numOfCells - 1;

$str = <STDIN>;
chomp($str);
my @cells = split(" ", $str);

$str = <STDIN>;
chomp($str);
my $acts = $str + 0;

for my $act (1...$acts) {
    $str = <STDIN>;
    chomp($str);
    my $actNum = $str + 0;
    my @moveLog = ();
    my $state = CURRENT;
    while ($state == CURRENT) {
        if ($actNum == $goal) {
            $state = FINISHED;
        }elsif($actNum < 1 || $goal < $actNum) {
            $state = DEAD;
        }else{
            my $moves = $cells[$actNum] + 0;
            if ($moves == 0) {
                $state = DEAD;
            }else{
                $actNum += $moves;
                foreach my $movedNum (@moveLog) {
                    if ($actNum == $movedNum) {
                        $state = DEAD;
                        last;
                    }
                }
                if ($state != DEAD) {
                    @moveLog = (@moveLog, $actNum);
                }
            }
        }
    }
    print ($state == FINISHED ? "Yes\n" : "No\n");
}
